### PR TITLE
add constraint info in PostgreSQLSchemaMetaDataLoader

### LINKED ISSUE #16086: Consider adding ConstraintMetaData to metadata to support constraint segment merge logic
## Feature Request

To support handling of foreign key constraints in issue #16050, ConstraintMetaData needs to be added to the metadata. 

Add `Map<String, Constraint MetaData> constraints` to TableMetaData:

```java
/**
 * Table meta data.
 */
@Getter
@EqualsAndHashCode
@ToString
public final class TableMetaData {
    
    private final String name;
    
    private final Map<String, ColumnMetaData> columns;
    
    private final Map<String, IndexMetaData> indexes;
    
    private final Map<String, ConstraintMetaData> constrains;
    
    private final List<String> primaryKeyColumns = new ArrayList<>();
}

/**
 * Constraint meta data.
 */
@RequiredArgsConstructor
@Getter
@EqualsAndHashCode
@ToString
public final class ConstraintMetaData {
    
    private final String name;
    
    private final String referencedTableName;
}
```

Then add query logic for constraints in MySQLTableMetaDataLoader.

```sql
SELECT CONSTRAINT_NAME, TABLE_NAME, COLUMN_NAME, REFERENCED_TABLE_NAME, REFERENCED_COLUMN_NAME FROM information_schema.KEY_COLUMN_USAGE
       WHERE TABLE_NAME IN ('bmsql_district_0') AND REFERENCED_TABLE_SCHEMA IS NOT NULL;
```

- [x] Support MySQL constraint meta data load
- [x] Support Postgresql constraint meta data load
- [ ] Support openGauss constraint meta data load
- [ ] Support Oracle constraint meta data load
- [ ] Support SQLServer constraint meta data load

### Is your feature request related to a problem?

issue #16050

### PR DESCRIPTION
**Pull request by lhyundeadsoul**
fix #16086

Changes proposed in this pull request:
- add constraint info in PostgreSQLSchemaMetaDataLoader
-
-

### CODE REVIEW

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/database/schema/loader/dialect/PostgreSQLSchemaMetaDataLoader.java [hunk 1]
```diff
@@ -45,48 +46,81 @@
  * Schema meta data loader for PostgreSQL.
  */
 public final class PostgreSQLSchemaMetaDataLoader implements DialectSchemaMetaDataLoader {
-    
+
     private static final String BASIC_TABLE_META_DATA_SQL = "SELECT table_name, column_name, ordinal_position, data_type, udt_name, column_default, table_schema"
             + " FROM information_schema.columns WHERE table_schema IN (%s)";
-
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/database/schema/loader/dialect/PostgreSQLSchemaMetaDataLoader.java:51]: Please keep the same indent with previous line.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/database/schema/loader/dialect/PostgreSQLSchemaMetaDataLoader.java [hunk 2]
```diff
@@ -45,48 +46,81 @@
...
         Collection<String> schemaNames = loadSchemaNames(dataSource, DatabaseTypeFactory.getInstance(getType()));
         Map<String, Multimap<String, IndexMetaData>> schemaIndexMetaDataMap = loadIndexMetaDataMap(dataSource, schemaNames);
         Map<String, Multimap<String, ColumnMetaData>> schemaColumnMetaDataMap = loadColumnMetaDataMap(dataSource, tables, schemaNames);
+        Map<String, Multimap<String, ConstraintMetaData>> schemaConstraintMetaDataMap = loadConstraintMetaDataMap(dataSource, schemaNames);
         Collection<SchemaMetaData> result = new LinkedList<>();
-        for (String each : schemaNames) {
-            Multimap<String, IndexMetaData> tableIndexMetaDataMap = schemaIndexMetaDataMap.getOrDefault(each, LinkedHashMultimap.create());
-            Multimap<String, ColumnMetaData> tableColumnMetaDataMap = schemaColumnMetaDataMap.getOrDefault(each, LinkedHashMultimap.create());
-            result.add(new SchemaMetaData(each, createTableMetaDataList(tableIndexMetaDataMap, tableColumnMetaDataMap)));
+        for (String schemaName : schemaNames) {
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/database/schema/loader/dialect/PostgreSQLSchemaMetaDataLoader.java:79]: Please rename schemaName to each.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/database/schema/loader/dialect/PostgreSQLSchemaMetaDataLoader.java [hunk 3]
```diff
@@ -45,48 +46,81 @@
...
+
+    private Map<String, Multimap<String, ConstraintMetaData>> loadConstraintMetaDataMap(final DataSource dataSource,
+                                                                                final Collection<String> schemaNames) throws SQLException {
+        Map<String, Multimap<String, ConstraintMetaData>> result = new LinkedHashMap<>();
+        try (Connection connection = dataSource.getConnection(); PreparedStatement preparedStatement = connection.prepareStatement(getConstraintKeyMetaDataSQL(schemaNames))) {
+            try (ResultSet resultSet = preparedStatement.executeQuery()) {
+                while (resultSet.next()) {
+                    String schemaName = resultSet.getString("table_schema");
+                    Multimap<String, ConstraintMetaData> constraintMetaData = result.computeIfAbsent(schemaName, key -> LinkedHashMultimap.create());
+
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/database/schema/loader/dialect/PostgreSQLSchemaMetaDataLoader.java:109]: Please remove this useless blank line.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/database/schema/loader/dialect/PostgreSQLSchemaMetaDataLoader.java (hunk 2 above)
Comment (lhyundeadsoul) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/database/schema/loader/dialect/PostgreSQLSchemaMetaDataLoader.java:79]: Is `schemaName` more explicit than `each` to reveal the object is here?

Comment (strongduanmu) [reply shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/database/schema/loader/dialect/PostgreSQLSchemaMetaDataLoader.java:79]: > Is `schemaName` more explicit than `each` to reveal the object is here?

You can read code conduct first.

**Comment by strongduanmu** [comment]
Hi @lhyundeadsoul, can you fix ci error?

**Comment by lhyundeadsoul** [comment]
> Hi @lhyundeadsoul, can you fix ci error?

sorry for the late reply.  It has been fixed. Thx.

File: shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/database/schema/loader/dialect/PostgreSQLSchemaMetaDataLoader.java [hunk 4]
```diff
@@ -107,6 +116,29 @@ private Collection<ViewMetaData> loadViewMetaData(final DataSource dataSource, f
...
     
+    private Map<String, Multimap<String, ConstraintMetaData>> loadConstraintMetaDataMap(final DataSource dataSource,
+                                                                                        final Collection<String> schemaNames) throws SQLException {
+        Map<String, Multimap<String, ConstraintMetaData>> result = new LinkedHashMap<>();
+        try (Connection connection = dataSource.getConnection(); PreparedStatement preparedStatement = connection.prepareStatement(getConstraintKeyMetaDataSQL(schemaNames))) {
+            try (ResultSet resultSet = preparedStatement.executeQuery()) {
+                while (resultSet.next()) {
+                    String schemaName = resultSet.getString("table_schema");
+                    Multimap<String, ConstraintMetaData> constraintMetaData = result.computeIfAbsent(schemaName, key -> LinkedHashMultimap.create());
+
```
Comment (strongduanmu) [inline shardingsphere-infra/shardingsphere-infra-common/src/main/java/org/apache/shardingsphere/infra/metadata/database/schema/loader/dialect/PostgreSQLSchemaMetaDataLoader.java:127]: Please remove useless blank line.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Review by strongduanmu — APPROVED** [review]
@lhyundeadsoul Good job, merged.

**Pull request merged**